ngraph/core/src/op/util/convert_color_i420_base.cpp

```diff
@@ -0,0 +1,218 @@
+// Copyright (C) 2018-2021 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "openvino/op/util/convert_color_i420_base.hpp"
+
+#include <memory>
+#include <ngraph/validation_util.hpp>
+
+#include "itt.hpp"
+#include "ngraph/runtime/reference/convert_color_nv12.hpp"
+#include "openvino/core/layout.hpp"
+
+static const size_t N_DIM = 0;
+static const size_t H_DIM = 1;
+static const size_t W_DIM = 2;
+static const size_t C_DIM = 3;
+
+ov::op::util::ConvertColorI420Base::ConvertColorI420Base(const Output<Node>& arg, ColorConversion format)
+    : Op({arg}),
+      m_format(format) {}
+
+ov::op::util::ConvertColorI420Base::ConvertColorI420Base(const Output<Node>& arg_y,
+                                                         const Output<Node>& arg_u,
+                                                         const Output<Node>& arg_v,
+                                                         ColorConversion format)
+    : Op({arg_y, arg_u, arg_v}),
+      m_format(format) {
+    constructor_validate_and_infer_types();
+}
+
+void ov::op::util::ConvertColorI420Base::validate_and_infer_types() {
+    NGRAPH_OP_SCOPE(v8_Convert_I420_Base_validate_and_infer_types);
+
+    NODE_VALIDATION_CHECK(this,
+                          get_input_size() == 1 || get_input_size() == 3,
+                          "I420 conversion shall have one or 3 inputs, but it is ",
+                          get_input_size());
+    auto single_plane = get_input_size() == 1;
+    auto y_type = get_input_element_type(0);
+    const auto& shape_y = get_input_partial_shape(0);
+    const auto one_channel_nhwc_shape =
+        PartialShape({Dimension::dynamic(), Dimension::dynamic(), Dimension::dynamic(), 1});
+    NODE_VALIDATION_CHECK(this,
+                          shape_y.compatible(one_channel_nhwc_shape),
+                          "Y input shall have 4 dimensions (N, H, W, C) with channels dimension equal to 1");
+    auto out_shape = shape_y;
+    auto out_type = y_type;
+    if (out_shape.rank().is_dynamic()) {
```

Comment:
why `shape_y` can have dynamic rank? I suppose it's always `4`?

Reply:
During build stage, PartialShape can be fully dynamic and be specified on 'inference' stage. This looks standard ngraph practice allowing fully dynamic shapes and element types on build stage.